File: android/src/main/java/org/ar/rtc/base/RtcMediaPlayer.kt
package org.ar.rtc.base

import org.ar.rtc.mediaplayer.ARMediaPlayerKit
import org.ar.rtc.rtmp.StreamingKit

class IRtcMediaPlayer{
    interface RtcMediaPlayerInterface{
        fun createInstance(callback: Callback)
        fun destory(callback: Callback)
        fun open(params: Map<String, *>,callback: Callback)
        fun play(callback: Callback)
        fun setView(params: Map<String, *>,callback: Callback)
        fun pause(callback: Callback)
        fun stop(callback: Callback)
        fun seek(params: Map<String, *>,callback: Callback)
        fun mute(params: Map<String, *>,callback: Callback)
        fun isMute(callback: Callback)
        fun adjustPlayoutVolume(params: Map<String, *>,callback: Callback)
        fun getPlayoutVolume(callback: Callback)
        fun getPlayPosition(callback: Callback)
        fun getDuration(callback: Callback)
        fun getState(callback: Callback)
        fun setRenderMode(params: Map<String, *>,callback: Callback)
        fun getStreamCount(callback: Callback)
        fun getStreamByIndex(callback: Callback)
    }
}

class RtcMediaPlayerManager:IRtcMediaPlayer.RtcMediaPlayerInterface{
    var player: ARMediaPlayerKit? =null

    override fun createInstance(callback: Callback) {
        player = ARMediaPlayerKit()
        callback.code(0)
    }

    override fun destory(callback: Callback) {
        player?.destroy()
        callback.code(0)
    }

    override fun open(params: Map<String, *>, callback: Callback) {
        callback.code(player?.open(params["url"] as String,(params["startPos"] as Int).toLong()))
    }

    override fun play(callback: Callback) {
        callback.code(player?.play())
    }

    override fun setView(params: Map<String, *>, callback: Callback) {
    }

    override fun pause(callback: Callback) {
        callback.code(player?.pause())
    }

    override fun stop(callback: Callback) {
        callback.code(player?.stop())
    }

    override fun seek(params: Map<String, *>, callback: Callback) {
        callback.code(player?.seek(params["position"] as Long))
    }

    override fun mute(params: Map<String, *>, callback: Callback) {
        callback.code(player?.mute(params["mute"] as Boolean))
    }

    override fun isMute(callback: Callback) {
        callback.resolve(player) { it.isMuted }
    }

    override fun adjustPlayoutVolume(params: Map<String, *>, callback: Callback) {
        callback.code(player?.adjustPlayoutVolume(params["vol"] as Int))
    }

    override fun getPlayoutVolume(callback: Callback) {
        callback.resolve(player) { it.playoutVolume }
    }

    override fun getPlayPosition(callback: Callback) {
        callback.resolve(player) { it.playPosition }
    }

    override fun getDuration(callback: Callback) {
        callback.resolve(player) { it.duration }
    }

    override fun getState(callback: Callback) {
        callback.resolve(player) { it.state }
    }

    override fun setRenderMode(params: Map<String, *>, callback: Callback) {
        callback.code(player?.setRenderMode(params["mode"] as Int))
    }

    override fun getStreamCount(callback: Callback) {
        callback.resolve(player) { it.streamCount }
    }

    override fun getStreamByIndex(callback: Callback) {
        callback.resolve(player) { it.streamByIndex }
    }

}

File: ios/Classes/Base/RtcMediaPlayer.swift
//
//  RtcMediaPlayer.swift
//  ar_rtc_engine
//
//  Created by 余生丶 on 2021/1/7.
//

import Foundation
import ARtcKit

protocol RtcMediaPlayerInterface {
    
    func createInstance(_ callback: Callback)
    
    func destory(_ callback: Callback)
    
    func open(_ params: NSDictionary,_ callback: Callback)
    
    func play(_ callback: Callback)
    
    func pause(_ callback: Callback)
    
    func stop(_ callback: Callback)
    
    func seek(_ params: NSDictionary, _ callback: Callback)
    
    func mute(_ params: NSDictionary, _ callback: Callback)
    
    func isMute(_ callback: Callback)
    
    func adjustPlayoutVolume(_ params: NSDictionary, _ callback: Callback)
    
    func getPlayoutVolume(_ callback: Callback)
    
    func getPlayPosition(_ callback: Callback)
    
    func getDuration(_ callback: Callback)
    
    func getState(_ callback: Callback)
    
    func setRenderMode(_ params: NSDictionary, _ callback: Callback)
    
    func getStreamCount(_ callback: Callback)
    
    func getStreamByIndex(_ params: NSDictionary, _ callback: Callback)
}

@objc
class RtcMediaPlayerManager: NSObject, RtcMediaPlayerInterface {
    
    private var emitter: (_ methodName: String, _ data: Dictionary<String, Any?>?) -> Void
    private(set) var player: ARMediaPlayer?
    private var delegate: RtcMediaPlayerEventHandler?

    init(_ emitter: @escaping (_ methodName: String, _ data: Dictionary<String, Any?>?) -> Void) {
        self.emitter = emitter
    }

    func Release() {
        player?.destroy()
        player = nil
        delegate = nil
    }
    
    @objc func createInstance(_ callback: Callback) {
        delegate = RtcMediaPlayerEventHandler() { [weak self] methodName, data in
            self?.emitter(methodName, data)
        }
        
        player = ARMediaPlayer.init(delegate: delegate!)
        callback.code(0)
    }
    
    @objc func destory(_ callback: Callback) {
        player?.destroy()
        callback.code(0)
    }
    
    @objc func open(_ params: NSDictionary, _ callback: Callback) {
        player?.open(params["url"] as! String, startPos: params["startPos"] as! Int)
        callback.code(0)
    }
    
    @objc func play(_ callback: Callback) {
        player?.play()
        callback.code(0)
    }
    
    @objc func pause(_ callback: Callback) {
        player?.pause()
        callback.code(0)
    }
    
    @objc func stop(_ callback: Callback) {
        player?.stop()
        callback.code(0)
    }
    
    @objc func seek(_ params: NSDictionary, _ callback: Callback) {
        player?.seek(toPosition: params["position"] as! Int)
        callback.code(0)
    }
    
    @objc func mute(_ params: NSDictionary, _ callback: Callback) {
        player?.mute(params["mute"] as! Bool)
        callback.code(0)
    }
    
    @objc func isMute(_ callback: Callback) {
        callback.resolve(player) { it in
            it.getMute()
        }
    }
    
    @objc func adjustPlayoutVolume(_ params: NSDictionary, _ callback: Callback) {
        callback.resolve(player) { (it) -> Any? in
            it.adjustVolume(Int32(params["vol"] as! Int))
        }
    }
    
    @objc func getPlayoutVolume(_ callback: Callback) {
        callback.resolve(player) { it in
            it.volume
        }
    }
    
    @objc func getPlayPosition(_ callback: Callback) {
        callback.resolve(player) { (it) -> Any? in
            it.getPosition()
        }
    }
    
    @objc func getDuration(_ callback: Callback) {
        callback.resolve(player) { (it) -> Any? in
            it.getDuration()
        }
    }
    
    @objc func getState(_ callback: Callback) {
        callback.resolve(player) { (it) -> Any? in
            it.state
        }
    }
    
    @objc func setRenderMode(_ params: NSDictionary, _ callback: Callback) {
        player?.setRenderMode(ARMediaPlayerRenderMode(rawValue: ARMediaPlayerRenderMode.RawValue(params["mode"] as! Int))!)
        callback.code(0)
    }
    
    @objc func getStreamCount(_ callback: Callback) {
        callback.resolve(player) { (it) -> Any? in
            it.getStreamCount()
        }
    }
    
    @objc func getStreamByIndex(_ params: NSDictionary, _ callback: Callback) {
        callback.resolve(player) { (it) -> Any? in
            it.getStreamBy(Int32(params["index"] as! Int))
        }
    }
}
